src/IO/Observer/ReductionActions.hpp

```diff
@@ -82,78 +83,220 @@ struct ContributeReductionData {
-                                     << reduction_names));
+            if (UNLIKELY(reduction_names_map->at(observation_id) !=
+                         reduction_names)) {
+              using ::operator<<;
+              ERROR("Reduction names differ at ObservationId "
+                    << observation_id << " with the expected names being "
+                    << reduction_names_map->at(observation_id)
+                    << " and the received names being " << reduction_names);
+            }
             reduction_data_map->operator[](observation_id)
                 .combine(std::move(reduction_data));
-            contribute_count++;
           }
 
+          ASSERT(observations_registered.find(ObservationIdRegistrationKey{
+                     observation_id}) != observations_registered.end(),
+                 "Couldn't find registration key "
+                     << ObservationIdRegistrationKey{observation_id}
+                     << " in the registered observers.");
           // Check if we have received all reduction data from the registered
           // elements. If so, we reduce to the local ObserverWriter nodegroup.
-          if (contribute_count == reduction_component_ids.size()) {
-            const auto node_id = Parallel::my_node();
+          if (UNLIKELY(contributed_array_ids.size() ==
+                       observations_registered
+                           .at(ObservationIdRegistrationKey{observation_id})
+                           .size())) {
             auto& local_writer = *Parallel::get_parallel_component<
                                       ObserverWriter<Metavariables>>(cache)
                                       .ckLocalBranch();
-            Parallel::threaded_action<ThreadedActions::WriteReductionData>(
-                local_writer, observation_id, subfile_name,
-                node_id == 0 ? std::move((*reduction_names_map)[observation_id])
-                             : std::vector<std::string>{},
+            Parallel::threaded_action<
+                ThreadedActions::CollectReductionDataOnNode>(
+                local_writer, observation_id,
+                ArrayComponentId{std::add_pointer_t<ParallelComponent>{nullptr},
+                                 Parallel::ArrayIndex<ArrayIndex>(array_index)},
+                subfile_name, (*reduction_names_map)[observation_id],
                 std::move((*reduction_data_map)[observation_id]));
             reduction_data_map->erase(observation_id);
             reduction_names_map->erase(observation_id);
             reduction_observers_contributed->erase(observation_id);
           }
         },
-        db::get<Tags::ReductionArrayComponentIds>(box));
+        db::get<Tags::ObservationsRegistered>(box));
   }
 };
 }  // namespace Actions
 
 namespace ThreadedActions {
+/*!
+ * \brief Gathers all the reduction data from all processing elements/cores on a
+ * node.
+ */
+struct CollectReductionDataOnNode {
```

Comment:
If memory serves correctly this was being done before. As for the lock, I'd like to replace this with either a simple lock-per-tag or a lock-free hash table (the latter being a rather complicated data structures to use correctly). It's already possible to write and contribute at the same time. I'm going to look into this a bit now. It would be reasonably easy to just change the tag type to a `pair<map, lock>` (especially with the NodeLock wrapper you wrote!). Ultimately, I think it only needs to be possible to write, contribute volume data, and contribute reduction data in parallel. Does that seem reasonable?

Reply:
Ah, the issue is that currently we are only allowed to call `db::mutate` in serial. While the underlying data can be mutated in parallel with various other DataBox operations (the DataBox guarantees pointer stability), we block all calls into the DataBox via a `bool` and a hard error in order to guarantee a persistent valid state. Basically, we can lock the node, retrieve the pointers, unlock the node, lock the data to be manipulated, manipulate the data, unlock the data to be manipulated. It's unclear to me what the cost of locking/unlocking is compared with just waiting for the operations.

Reply:
Ah, understood. I think given that the possible gains are not easy to reason about, I'm happy with keeping the current implementation if only because it is simpler to read and maintain without the extra locking operations.

Reply:
Well I already had it implemented by the time you added this comment :shrug: now whether there is any performance improvement, :shrug:

Reply:
heh, too fast for me -- Well, I'm certainly happy with the more granular locking procedure now that you've done all the labor of putting it in :) .

Reply:
:+1: It was surprisingly not too bad to do. Basically two new tags and some non-owning pointers in placing :shrug:

I'm going to write a dev guide for the observer code tomorrow and then push the fixups. There'll still be some maintenance work (@kidder's renames and moving files and actions around), but I'm going to defer that